[iOS] Fixed Shell Navigating event showing same current and target values

Vignesh-SF3580 opened the pull request:
### Issue Detail

In the Shell Navigating event, the Current and Target locations are the same.

### Root Cause

The route comparison logic did not properly check the navigation state, leading to redundant route values in routeStack.

### Description of Change

Updated the route validation logic to correctly compare and update routeStack. If the new route matches the existing one at the expected position, redundant elements in routeStack are removed to maintain an accurate navigation state.

### Tested the behaviour in the following platforms
 
- [x] Android
- [x] Windows
- [x] iOS
- [x] Mac

### Issues Fixed
 
Fixes https://github.com/dotnet/maui/issues/25599

### Screenshots

| Before Issue Fix | After Issue Fix |
|----------|----------|
| <video width="300" height="600"  src="https://github.com/user-attachments/assets/2706b3c6-f289-4eb7-9f4a-9475898bff87"> | <video width="300" height="600"  src="https://github.com/user-attachments/assets/86c72f4c-f6b3-4a7c-b39c-2245fa93d885">) |

Vignesh-SF3580 committed:
Fixed - 25599 : OnNavigating wrong target when tapping the same tab

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Controls/src/Core/Compatibility/Handlers/Shell/iOS/ShellItemRenderer.cs:
```diff
@@ -113,9 +113,9 @@ public override void ViewDidLoad()
+				bool accept = false;
 				var r = RendererForViewController(viewController);
-				if (r != null)
+				if (r is not null && r != SelectedViewController)
```
Could include an UITest?

Vignesh-SF3580 replied:
I have added the test case. Could you please review it and let me know if you have any concerns?

Vignesh-SF3580 committed:
Added testcasees.

rmarinho commented:
/rebase

rmarinho commented:
/azp run

jsuarezruiz commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue25599.cs:
```diff
@@ -0,0 +1,27 @@
+﻿#if ANDROID // Facing a full exception when clicking on the already selected tab(App.Tap()) in Windows, iOS, and macOS, so excluded those platforms.
```
Could you share the exception? Want to determinate if is related with a possible bug on Android or if we need to review something at testing stuff.

Vignesh-SF3580 replied:
I have revalidated the test cases once more with the latest source and I can confirm that it is now possible to click on the already selected tab using the tab title on Windows and AutomationId on iOS and Mac.
 
I have modified the test cases to run successfully in all platforms. Could you please review and let me know if you have any concerns?

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/rebase

jsuarezruiz commented:
/azp run

PureWeen commented:
/azp run

PureWeen reviewed (changes requested):
1) Can you split these changes into two PRs?
2) I think the implementation on iOS isn't addressing the issue. The problem the user is reporting is that the navigating event isn't firing when they reclick on the Tab. Ideally we don't disable this behavior but we instead route it through the Shell.Navigating features so that users can enable/disable this behavior

Vignesh-SF3580 committed:
Removed windows fix and created new PR for windows fix.

Vignesh-SF3580 commented:
> 1. Can you split these changes into two PRs?
> 2. I think the implementation on iOS isn't addressing the issue. The problem the user is reporting is that the navigating event isn't firing when they reclick on the Tab. Ideally we don't disable this behavior but we instead route it through the Shell.Navigating features so that users can enable/disable this behavior

1. I have divided the changes into two separate PRs: [25749](https://github.com/dotnet/maui/pull/25749) and [27197](https://github.com/dotnet/maui/pull/27197).
2. Regarding the reported issue: The user highlighted that clicking on an already selected tab triggers the  **Navigating** event with the same navigation parameters. This behavior is inconsistent across platforms.

**Current Behavior:**
* **iOS and macOS:** Clicking on an already selected tab triggers both the **Navigating** and **Navigated** events, and navigation occurs.
* **Android:** Clicking on an already selected tab does not trigger the **Navigating** or **Navigated** events, and no navigation occurs.
* **Windows:** Switching between tabs does not trigger the Navigating or Navigated events. To address this, I have implemented logic on Windows to ensure these events are triggered. ([27197](https://github.com/dotnet/maui/pull/27197))

**Behavior After the Fix:**
* Clicking on an already selected tab will no longer trigger the Navigating event, and no navigation will occur.
* The Navigating event will only be triggered when switching to a different (unselected) tab.

**Example Scenario:**
Consider a tabbed application with three tabs: **Home**, **Profile**, and **Settings**.
* When the user selects the **Home** tab, the Navigating event is triggered, and the application navigates to the default Home view.
* From the Home view, the user navigates to another page (e.g., AnotherPage) by clicking a button. If the user then clicks the **Home** tab again (which is already selected), the Navigating event will **not** be triggered, and no navigation will occur. Returning to the previous page can only be done using the back **navigation** button.
* However, if the user switches to the **Profile** tab, the Navigating event will be triggered, and the application will navigate to the Profile view.
 
**Note:** I implemented this based on the behavior observed in Android.

PureWeen reviewed (changes requested):
To clarify on the after behavior @Vignesh-SF3580 

We still want navigation to happen on iOS when they reselect the tab but we want the NAvigating event to fire so that users have the option to cancel

Can you also add a test?

Vignesh-SF3580 committed:
Fixed-25599 : OnNavigating wrong target when tapping the same tab

Vignesh-SF3580 committed:
Revert "Added testcasees."

This reverts commit 8df87c72609cf42748398b8b86664bc910f77873.

Vignesh-SF3580 committed:
Revert "Fixed - 25599 : OnNavigating wrong target when tapping the same tab"

This reverts commit 366c51d7ce5ad543436a9c53165036015b8ed576.

Vignesh-SF3580 committed:
removed unwanted changes

Vignesh-SF3580 committed:
added test and addressed copilot comments
